[임루트] 1단계 - 자동 미션 제출합니다.

ttkmw opened the pull request:
안녕하세요, 임루트입니다.

**2/20 18:00**
최대한 완성하려 푸시하려고 했으나, 아직 많이 부족하여 완성은 못하고 푸시하였습니다.
죄송합니다.
미션을 시간내에 완료하기 위해 리뷰요청은 보내지만, 추가적으로 푸시를 하겠습니다.

-------------------------------------------------------------------------
**2/20 22:00**
현재 기능 구현은 완료하였습니다.
다만, 프로그래밍 요구사항은 아직 지키지 못했습니다.
구체적으로, 아직 아래 두가지를 지키지 못하였습니다.
_규칙 3: 모든 원시값과 문자열을 포장한다.
규칙 8: 일급 콜렉션을 쓴다._

규칙을 지키도록 리팩토링 하겠습니다.
**_완료가 되면, 풀리퀘스트 메시지에 완료가 됐다고 수정하도록 하겠습니다._**

-------------------------------------------------------------------------
**2/21 18:27**
페어와 리팩토링 할 부분을 정리 후, 일부를 리팩토링 했습니다.
그런데 그 과정에서, 로또게임을 구동시키는 부분의 코드가 깨져, 깨지는 부분을 주석처리 해놓은 상태입니다.
이후 다시 수정하도록 하겠습니다.

-------------------------------------------------------------------------
**2/24 18:04**
드디어 리팩토링을 완료했습니다!
확인 부탁드립니다.
피드백 더 주시면, 감사한 마음으로 추가수정하도록 하겠습니다!


감사합니다.

sonypark committed:
docs: 미션개요, 구현상세, 기능목록 작성

sonypark committed:
critical: 도메인, 뷰 설계

sonypark committed:
docs: 클래스명 변경으로 인한 문서 수정: 발급 -> 생성

sonypark committed:
feat: 임의로 1부터 45까지의 숫자를 6개 생성하여 리스트로 만든다

sonypark committed:
refac: 임의로 1부터 45까지의 숫자를 6개 생성하여 리스트로 만든다

sonypark committed:
docs: '로또 번호 중복 허용 불가' 기능 추가

sonypark committed:
feat: 로또 숫자 생성 시 6개의 숫자는 중복되지 않는다

sonypark committed:
refac: 테스트 메서드명 변경: generate -> 로또숫자_생성_갯수

sonypark committed:
feat: 구매 금액을 생성할 로또 개수로 변환한다

sonypark committed:
feat: 로또 개수만큼 로또를 생성한다

sonypark committed:
refac: 로또 개수만큼 로또를 생성한다

sonypark committed:
feat: 생성된 로또와 지난 주 당첨번호를 비교한다

sonypark committed:
refac: LottosGenerator의 generate() 인자값 변경

Lotto 생성자 변경에 따라 LottosGenerator의 generate() 인자값을 변경함

sonypark committed:
refac: Lotto의 match() 함수를 matchWithWinningNumbers()로 변경

이전에 bonusBall을 한번에 비교하려고 잘못 생각해서 변경

sonypark committed:
feat: 로또번호와 보너스볼 비교(일치 여부 확인)

sonypark committed:
docs: 당첨 금액 규칙 중 명칭 변경

기존에 일치한 숫자로 표현한 것을 등수로 변경

sonypark committed:
refac: LottoNumbersGenerator의 generate() 로직 변경

기존에 로직 반영이 잘 안된 부분을 수정

sonypark committed:
feat: 당첨 금액 산출 규칙에 따라 순위를 산출한다.

sonypark committed:
feat: 당첨 금액 산출 규칙에 따라 당첨 금액을 산출한다

sonypark committed:
feat: 수익률을 계산한다

sonypark committed:
feat: 총 당첨 금액을 계산한다

sonypark committed:
feat: 로또 게임을 구동한다

sonypark committed:
ing: 출력 부분 진행중입니다.

sonypark committed:
chore: 제출 전 급하게 구현한 출력 로직 삭제

sonypark committed:
feat: 수익률을 출력한다

sonypark committed:
feat: 결과를 담는 객체를 만들고, 결과를 추가

sonypark committed:
feat: 랭크별 당첨된 티켓 수 조회

sonypark committed:
feat: 당첨 통계 출력

sonypark committed:
refac: 당첨 결과를 계산

sonypark committed:
refac: 로또 게임 구동

sonypark committed:
test: StringUtils 테스트 추가

sonypark committed:
refac: StringUtils 안의 변수명 변경(STANDARD -> DELIMITER)

sonypark committed:
refac: LottoNumbersGenerator 내 함수 분리

이외에 random에 대한 학습 테스트, 접근 제한자 수정 등의 작업이 있음

sonypark committed:
refac: Rank의 함수명 변경(isExists -> isValid)

sonypark committed:
refac: lottoNumbers의 타입 변환(List -> Set)

이외에 convertMoneyToLottosSize() 함수의 위치 변경, lottoNumbersString의
네이밍 변경

sonypark committed:
refac: EarningRateCalculator 상수명 변경

dave915 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,34 @@
+        return (int)lottoNumbers.stream()
+            .parallel()
+            .filter(lottoNumber -> winningNumbers.contains(lottoNumber))
+            .count();
```
아래 글을 읽어보면 좋을 것 같아요. 
https://sejoung.github.io/2019/02/2019-02-12-Use_caution_when_making_streams_parallel/

또 경험한 바로는 ParallelStream은 `Common Thread Pool` 을 사용하기 때문에 병렬로 처리 하는 작업시간이 오래 걸릴 경우 Api같은 Application에서 사용한다면 Request가 밀리는 현상이 있었습니다. 기본적으로 `Common Thread Pool`은 Cpu코어수 만큼 Thread를 생성 하고 각 Request가 모두 `Common Thread Pool`를 사용하기 때문에 다량으로 Request가 들어왔을때 `Common Thread Pool`이 앞서 들어온 작업들을 모두 처리 할 때까지 기달려야하기 때문이에요. 
물론 `Common Thread Pool`의 Thread갯수를 조절 할 수 있지만 기본적으로 `정말 해당 로직이 병렬로 처리되어야 할까?`부터 고민하면 좋을 것 같네요 :)

ttkmw replied:
와...좋은 자료 감사합니다!!
처음 읽었을 땐 잘 이해가 안됐는데 몇 번 읽어보니 조금 알 것 같습니다.
성능 관련해서는 적재적소에 사용하는게 굉장히 어렵네요...
감사합니다!

dave915 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,34 @@
+            .map(lottoNumber -> lottoNumber.toString())
+            .collect(Collectors.toList());
+
+        return "[" + String.join(",", stringifiedLottoNumbers) + "]";
```
[로또번호] 로 출력 형식을 만드는 것은 View의 영역일 것 같아요.
View에 따라서 {로또번호} 형식으로 출력 할 수 있으니까요.

ttkmw replied:
그렇네요! getter로 lottoNumbers를 구할 수 있게 하고 뷰에서 로직을 처리하도록 했습니다!

dave915 commented on src/main/java/lotto/domain/LottoNumbersGenerator.java:
```diff
@@ -0,0 +1,43 @@
+            lottoNumbers = fillUpLottoNumber(lottoNumbers, lottoNumber);
+        }
+        return lottoNumbers;
+    }
```
1. List<Integer>로 둘 경우, 메모리 문제나 성능 상 문제가 생기지 않을까 걱정이 됐습니다.
lottoNumbers객체는 ArrayList의 참조값을 갖고 있고 fillUpLottoNumber 내부에서 계속해서 ArrayList를 생성하는 것이아니라 parameter로 들어온 참조값을 그대로 return해주기 때문에 메모리 문제는 없을 것 같아요.

2. 가독성 측면에서 void가 나은 지, List<Integer>가 나은 지 헷갈렸습니다.
현재 코드상으로 봤을땐 가독성보단 해당 메소드가 lottoNumbers의 참조값을 받아 내부에서 작업하기 때문에 굳이 List<Integer>로 줄 필요는 없어보입니다. 
하지만 로또 번호를 채우는 부분을 추상화 한다고 했을때는 return값이 있는 메소드로 만드는게 더 유연한 추상화가 가능하지 않을까 싶기도 해요. 구현체 내부에서 새롭게 생성된 list를 return 할수도 있으니까요.

그리고 메소드의 return값과 별개로 return값이 있어도 아래처럼 값을 받지 않으면 상관 없지 않을까요? 
```
    static List<Integer> generate() {
        List<Integer> lottoNumbers = new ArrayList<>();
        while (isFulfilled(lottoNumbers)) {
            int lottoNumber = pickLottoNumber();
            fillUpLottoNumber(lottoNumbers, lottoNumber);
        }
        return lottoNumbers;
    }
```

ttkmw replied:
친절한 설명 감사드립니다...!
이해가 잘 됐습니다
감사합니다!

dave915 commented on src/main/java/lotto/domain/LottosGenerator.java:
```diff
@@ -0,0 +1,20 @@
+
+    static List<Lotto> generate(int lottosSize, List<Set<Integer>> lottoNumbersList) {
+        //질문: final을 붙이는 것이 의미가 있는 것인 지 궁금합니다. 제 생각에는 이후에 초기화를 더 하지 못하게 막음으로써 메모리 누수를 막을 수 있다고 생각하는데요. 어떻게 생각하시나요?
+        final List<Lotto> lottos = new ArrayList<>();
```
저도 아직 final을 명시적으로 붙이는 것에 습관 들이지 못했는데요.. 
다른 개발자와 함께 코딩하는 프로젝트에서 명시적으로 불변 객체인것을 나타낼수 있기 때문에 붙이는게 좋다고 생각합니다. 
메모리 누수도 있지만 명시적으로 표현해주는거죠 :)

ttkmw replied:
그렇군요 답변 감사합니다!

dave915 commented on src/main/java/lotto/domain/LottosGenerator.java:
```diff
@@ -0,0 +1,20 @@
+            lottos.add(lotto);
+        }
+
+        return lottos;
```
lottos를 그대로 return하면 ArrayList의 참조값을 return하기 때문에 list내부가 변경이 가능해집니다 😭

ttkmw replied:
변경하지 못하도록 수정했습니다!

dave915 commented on src/test/java/lotto/domain/RankTest.java:
```diff
@@ -0,0 +1,73 @@
+        //given
+        int invalidWinningNumber = 2;
+        //질문: 제 생각에는 다른 사람이 보기에 invalidWinningNumber가 어떤 의미인지 알기 어려울 것 같아, rank에 없는 값이라는 것을 알려주고 싶어 아래 assertTrue를 넣었습니다. 이게 필요 없는 것일 지 궁금합니다.
+        assertTrue(Arrays.stream(Rank.values()).allMatch(rank -> rank.getMatchNumber() != invalidWinningNumber));
```
어차피 테스트코드이기 때문에 필요 없는것은 없습니다 ㅎㅎ
개인적인 생각으로는 주석으로 알려줘도 되지 않을까 싶네요.

ttkmw replied:
넵 주석으로 알려주도록 수정했습니다!

dave915 commented on src/test/java/lotto/domain/RankTest.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Test
+    void 당첨된_갯수에_해당하는_순위가_없음() {
```
개인적으로 테스트 메소드 명은 영문으로 하고 자세한 설명은 @DisplayName을 이용하는 것을 선호합니다 :)

ttkmw replied:
넵 영문화 후 DisplayName을 활용해봤습니다!

dave915 commented on src/main/java/lotto/domain/LottoGame.java:
```diff
@@ -0,0 +1,82 @@
+import lotto.view.InputView;
+import lotto.view.OutputView;
+
+public class LottoGame {
```
모두 주석처리 해 놓으신 이유가 있을까요?

ttkmw replied:
코드가 깨져 그렇게 해놨었는데, 이제 작동 가능하도록 수정했습니다!ㅎㅎ

dave915 commented on src/main/java/lotto/domain/TotalWinningMoneyCalculator.java:
```diff
@@ -0,0 +1,10 @@
+public class TotalWinningMoneyCalculator {
+
+    static int calculate(List<Integer> winningMoneys) {
+        return winningMoneys.stream().mapToInt(Integer::intValue).sum();
```
객체지향 생활체조 `규칙 4: 한 줄에 한 점만 사용`를 읽어보세요~!
https://developerfarm.wordpress.com/2012/01/30/object_calisthenics_5/

ttkmw replied:
넵 리팩토링 중 이 부분이 아예 없어져서, 자연스럽게 수정됐습니다!

dave915 reviewed (changes requested):
임루트 안녕하세요! 리뷰를 맡게된 데이브입니다.
구현 잘 해주셨네요 ㅎㅎ 👍
말씀하신대로 LottoGame 주석 해 놓으신 부분 해결 하고 다시 PR주세요! 
질문 주신 부분에 대해 답변드렸어요 ㅎㅎ 확인부탁드리고, 궁금하신 부분 있으시면 언제든 DM주세요!

sonypark committed:
refac: 피드백 반영

1. 변수명 변경: PERCENTILE -> PERCENT
2. parallel 삭제
3. 클래스 변경 및 삭제: LottoComparator,LottoNumberConfig,TotalWinningMoneyCalculator 삭제 및 로직 이동
4. LottoNumbersGeneratorTest: 로또 숫자 유효범위 테스트 추가

sonypark committed:
refac: 피드백 반영 : Random객체 사용 -> Collections.shuffle() 사용

sonypark committed:
refac: 피드백 반영 - Rank의 enum값 자체가 스스로 자신에 해당하는지 판단하도록 함

sonypark committed:
refac: 원시값 포장 - Money

sonypark committed:
refac: 원시값 포장 - LottoNumber

sonypark committed:
refac: 일급 콜렉션 적용: LottoNumberBox

sonypark committed:
refac: LottoNumbersGenerator -> LottosGenerator 로 변경

sonypark committed:
refac: 당첨금액 총합과 수익률을 가지는 result 객체를 만듦

sonypark committed:
refac: 로또 게임 구동

sonypark committed:
docs: README에 사용 방법 추가

ttkmw committed:
refac: 피드백 반영

dave915 commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -0,0 +1,41 @@
+        printEarningRate(result.getEarningRate());
+    }
+
+    private static void printStatistics(WinningRanks winningRanks) {
+        Map<Rank, Integer> winningResults = winningRanks.getWinningRanks();
+        for (Rank rank : winningResults.keySet()) {
+            System.out.println(
+                    String.format("%d개 일치(%d원)- %d개", rank.getMatchNumber(), rank.calculateWinningMoney().getValue(),
+                            winningResults.get(rank)));
+        }
+    }
```
String.format이 길어지면 따로 변수로 분리하면 좋을 것 같아요 :) 
```
    private static void printStatistics(WinningRanks winningRanks) {
        Map<Rank, Integer> winningResults = winningRanks.getWinningRanks();
        for (Rank rank : winningResults.keySet()) {
            String resultMessage = String.format("%d개 일치(%d원)- %d개", rank.getMatchNumber(), 
                    rank.calculateWinningMoney().getValue(), winningResults.get(rank));
            System.out.println(resultMessage);
        }
    }
```

ttkmw replied:
그렇네요 훨씬 깔끔한 것 같습니디ㅏ!

dave915 commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,29 @@
+    private static final Scanner scanner = new Scanner(System.in);
+
+    public static Money inputPurchaseAmount() {
+        System.out.println("구입금액을 입력해주세요");
```
메세지를 상수로 올리면 좋을 것 같아요 :)

ttkmw replied:
@dave915
혹시 메시지를 상수로 뺐을 때의 장점을 알려주실 수 있을까요?
제 생각엔 메시지를 상수로 빼다보면,
너무 많은 상수가 생기고
그러면 가독성이 안좋을 것 같아서요.
그리고 해당 메시지가 재사용 가능성이 많지 않다고 생각하여, 상수로 뺐을 때의 장점이 궁금해졌습니다!

dave915 commented on src/test/java/lotto/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,27 @@
+
+    @ParameterizedTest
+    @ValueSource(ints = {0, 46})
+    @DisplayName("값이 비정상 범위에 있을 때 로또 번호 생성")
```
@DisplayName 👍

dave915 commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -0,0 +1,41 @@
+        System.out.println(String.format("%d개를 구매했습니다.", lottosSize));
+    }
+
+    public static void printLotto(Lotto lotto) {
+        List<String> stringifiedLottoNumbers = lotto.getLottoNumbers().stream()
+                .map(LottoNumber::toString)
+                .collect(Collectors.toList());
+
+        System.out.println("[" + String.join(",", stringifiedLottoNumbers) + "]");
+    }
```
아래처럼 쓸 수 있어요 😎
```
    public static void printLotto(Lotto lotto) {
        String lottoString = lotto.getLottoNumbers().stream()
                .map(LottoNumber::toString)
                .collect(Collectors.joining(",", "[", "]"));

        System.out.println(lottoString);
    }
```

ttkmw replied:
훨씬 깔끔하네요! 감사합니다.

dave915 commented on src/main/java/lotto/utils/StringUtils.java:
```diff
@@ -0,0 +1,20 @@
+
+    public static List<Integer> parseWithDelimiter(String input) {
+        return Arrays.stream(input.split(DELIMITER))
+            .mapToInt(number -> Integer.parseInt(number))
```
람다식의 실행코드가 한줄일때 람다식의 파라미터 갯수 와 실행되는 메소드의 파라미터 갯수가 같으면 메소드 참조 형식으로 쓸 수 있어요.
```suggestion
            .mapToInt(Integer::parseInt)
```

ttkmw replied:
넵 심지어 인텔리제이가 힌트를 줬는데도 지나쳤더라고요...ㅜㅜ 반성합니다.
알려주셔서 감사합니다!

dave915 commented on src/main/java/lotto/domain/WinningRanks.java:
```diff
@@ -0,0 +1,42 @@
+    /* winningTicketSize은 각 Rank별 당첨된 티켓 수를 의미한다. */
+    Map<Rank, Integer> addWinningRanks(Rank rank) {
+        Integer winningTicketSize = winningRanks.get(rank);
+        if (winningTicketSize == null) {
```
```suggestion
        if (Objects.isNull(winningTicketSize)) {
```

저는 null check를 할때 Objects클래스에서 제공해주는 메소드를 사용하는 편이에요.
https://multifrontgarden.tistory.com/205

ttkmw replied:
그렇군요 감사합니다!

dave915 commented on src/main/java/lotto/domain/LottoNumberBox.java:
```diff
@@ -0,0 +1,19 @@
+    private static final int LOTTO_MAX_NUMBER = 45;
+    private static final int LOTTO_MIN_NUMBER = 1;
+    private static final int ONE = 1;
+    private static final List<LottoNumber> LOTTO_NUMBERS = IntStream.range(LOTTO_MIN_NUMBER, LOTTO_MAX_NUMBER + ONE)
```
```suggestion
    private static final List<LottoNumber> LOTTO_NUMBERS;
    
    static {
        LOTTO_NUMBERS = IntStream.range(LOTTO_MIN_NUMBER, LOTTO_MAX_NUMBER + ONE)
                .mapToObj(LottoNumber::new)
                .collect(Collectors.toList());
    }
```
static변수의 초기화 값에 로직이 필요하다면 static 블록을 사용해보는건 어떨까요? 
https://uoonleen.tistory.com/6

ttkmw replied:
감사합니다 수정했습니다!

dave915 reviewed (approved):
피드백 반영한 부분 확인했어요! 👍
다음 Step진행을 위해 여기서 머지할게요.
코멘트 몇개 남겼으니 확인해보세요 :)
